[RFC] Runtime env agent: standlone python http server

rynewang opened the pull request:
## Why are these changes needed?

1. broke runtime env agent from dashboard agent into a standalone process.
2. The process uses http, not grpc.


<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

Part of issue #35472. Competition PR #36667. 

## Approach

1. (common with #36667) remove the runtime env agent from dashboard agent
2. (common with #36667) remove the dashboard agent registration grpc call because it no longer need to notify node manager its address. Correspondingly, remove node manager's dashboard agent treating callbacks (but keeps dashboard agent creation).
3. moves the python runtime env agent code into a standalone package in `ray/_private/runtime_env/agent`.
4. Adds `aiohttp` as a private vendored library in `setup.py`. In the wheel it's in `ray/_private/runtime_env/agent/thirdparty_files`.
5. Use `aiohttp` to implement a http server that keeps a runtime env agent, recv calls, parse/unparse protos, and invoke the agent methods.
6. adds a new `RuntimeEnvManagerClient` to node manager, which handles calls from worker pool. Upon call it forwards calls via HTTP to the server in step 5.
7. (todo) spin up the server in node manager ctor, and make it fate share with raylet itself.
9. (todo) clean up: update gitignore, pass parameters properly.

rynewang commented:
RFC v2.

- `agent_manager.h` is now "general purpose" and is used to spin up a dashboard agent, and a runtime env agent separately.
- renamed most `agent_*` into `dashboard_agent_*` and added corresponding `runtime_env_agent_*`. This applies to node_manager and python `services.py`, and `consts.py`.
- runtime env agent (`main.py`) is now a standalone python HTTP server.
- runtime env agent has 2 endpoints, corresponding to the previous 2 methods used by raylet. Both endpoints are POST, and carries `application/octet-stream` protobuf binaries.
- the behavior should be exactly as before, including all plugins.
- I tested `ray.init(runtime_env={"pip":["emoji"]})` and it works.
- runtime env agent's all dependencies (only aiohttp and virtualenv, but with their transitive deps) are now vendored. sum of all vendored libs == 13 MB. (see below for a list)
- runtime env agent fate-shares with raylet. dashboard agent now no longer fate-shares with raylet. i.e. if raylet dies, both agents dies. If runtime env agent dies, raylet dies. But if dashboard agent dies, raylet does NOT die.

TODOs:

- The third method, `GetRuntimeEnvsInfo` is not implemented. It's used by `state_manager.py`.
- Python tests are expected to pass, but C++ tests need redo.

```
(venv) (base) ruiyangwang@ryw-Q9LXJFG3Q4 ray % cd python/ray/_private/runtime_env/agent/thirdparty_files
(venv) (base) ruiyangwang@ryw-Q9LXJFG3Q4 thirdparty_files % du -sh .
 13M	.
(venv) (base) ruiyangwang@ryw-Q9LXJFG3Q4 thirdparty_files % ls
aiohttp					attr					distlib					idna					virtualenv
aiohttp-3.8.4.dist-info			attrs					distlib-0.3.6.dist-info			idna-3.4.dist-info			virtualenv-20.21.0.dist-info
aiosignal				attrs-23.1.0.dist-info			filelock				multidict				yarl
aiosignal-1.3.1.dist-info		bin					filelock-3.12.2.dist-info		multidict-6.0.4.dist-info		yarl-1.9.2.dist-info
async_timeout				charset_normalizer			frozenlist				platformdirs
async_timeout-4.0.2.dist-info		charset_normalizer-3.1.0.dist-info	frozenlist-1.3.3.dist-info		platformdirs-3.8.0.dist-info
```

aslonnie reviewed (commented):
the change to `ci/docker/base.test.Dockerfile` looks ok, given that there are already many stuff inthere.. 

will leave others to review the meat first; please re-request review when you give approvals from other related people already.

rynewang commented:
RFC v3

Not much from v2, but with more fixes.

- Used boost beast http client. It is async (works on io_context), so we don't need to think about making another thread.
- Removed cpr. This means we no require libcurl in development and in runtime. Also it should work out-of-the-box in windows.
- Removed virtualenv from vendored dependencies. It's still a dependency for `pip` runtime env plugin, but we don't vendor it now.

Smoke tested with `python -c 'import ray; ray.init(runtime_env={"pip":["emoji"]})'` and the logs show up the delayed creation of envs.

It is ripe to review. But we still have todo:

- Fix C++ tests to make CI happy.
- (optional, in another PR) use HTTPS for the agent?

scv119 commented:
@rynewang you need to set the assignee so it will show up in our review list :)

scv119 commented on python/ray/_private/runtime_env/agent/main.py:
```diff
@@ -0,0 +1,186 @@
+)
+
+
+def import_libs():
```
should we add test for this?

scv119 commented on python/ray/_private/runtime_env/agent/main.py:
```diff
@@ -0,0 +1,186 @@
+import runtime_env_consts
+from runtime_env_agent import RuntimeEnvAgent
+from aiohttp import web
+
```
nit: extra line

rkooo567 commented:
Is it possible to split the PR? I think we can make it as 2 PRs (1 and 2 separately)?

rynewang commented:
I believe I have fixed all new failing tests. The only outstanding ones are `//python/ray/tests:test_gcs_fault_tolerance` which is actually flaky on my ci repro aws instance; and `//java:all_tests` which should not have runtime env issues.

rynewang commented:
We are splitting this PR into pieces and gradually submitting them. See #37585
